### PR TITLE
Ft add overall statistic

### PR DESCRIPTION
**Pull request by esirK**
## Description

Adds Overall statistics

Fixes # (issue)

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## Screenshots
# Sheet 5
Screenshot 2021-03-17 at 10 25 40

# Sheet 11
Screenshot 2021-03-17 at 10 26 48

# Sheet 13
Screenshot 2021-03-17 at 10 27 11

# Sheet 28 
Screenshot 2021-03-17 at 10 27 42

# Sheet 30
Screenshot 2021-03-17 at 10 28 02

# Sheet 38 
Screenshot 2021-03-17 at 10 28 27

# Sheet 41
Screenshot 2021-03-17 at 10 28 52

# Sheet 47 
Screenshot 2021-03-17 at 10 29 12

# Sheet 49
Screenshot 2021-03-17 at 10 29 53

# Sheet 50
Screenshot 2021-03-17 at 10 30 24

# Sheet 51
Screenshot 2021-03-17 at 10 31 04

# Sheet 52
Screenshot 2021-03-17 at 10 31 25

# Sheet 62
Screenshot 2021-03-17 at 10 31 51

# Sheet 63
Screenshot 2021-03-17 at 10 32 11

# Sheet 65
Screenshot 2021-03-17 at 10 32 31

# Sheet 68b
Screenshot 2021-03-17 at 10 32 53

# Sheet 100
Screenshot 2021-03-17 at 10 33 16

# Sheet 101
Screenshot 2021-03-17 at 10 33 39

# Sheet 102
Screenshot 2021-03-17 at 10 34 02

# Sheet 103
Screenshot 2021-03-17 at 10 34 23

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation

### CODE REVIEW

File: reports/report_builder.py [hunk 1]
```diff
@@ -648,12 +648,28 @@ def ws_05(self, ws):
                     counts.update({(r['sex'], TOPIC_GROUPS[r['topic']]): r['n']})
 
             counts_list.append(secondary_counts)
+        overall_column = ws.dim_colmax
         self.tabulate_secondary_cols(ws, counts_list[0], self.male_female, MAJOR_TOPICS, row_perc=True)
         c = ws.dim_colmax + 2
+        overall_row = ws.dim_rowmax+2
+        for media_type in counts_list[0]:
+            overall_total_dm = sum([counts_list[0][media_type][x] for x in counts_list[0][media_type]])
```
Comment (kilemensi) [inline reports/report_builder.py]:
```suggestion
            overall_total_tm = sum([counts_list[0][media_type][x] for x in counts_list[0][media_type]])
```

File: reports/report_builder.py [hunk 2]
```diff
@@ -648,12 +648,28 @@ def ws_05(self, ws):
...
+            female_values = [x for x in counts_list[0][media_type] if x[0]==self.female_ids[0]]
+            overall_tm_female_value = sum([counts_list[0][media_type][x] for x in female_values]) / overall_total_dm
```
Comment (kilemensi) [inline reports/report_builder.py]:
```suggestion
            overall_tm_female_value = sum([counts_list[0][media_type][x] for x in female_values]) / overall_total_tm
```

Mixing `tm` and `dm` I think

File: reports/report_builder.py [hunk 3]
```diff
@@ -791,9 +807,16 @@ def ws_11(self, ws):
 
                 for r in rows:
                     counts.update({(r['equality_rights'], TOPIC_GROUPS[r['topic']]): r['n']})
+        overall_column = ws.dim_colmax
+        yes_no_sum = sum(counts.values())
+        yes_sum = sum([counts[x] for x in counts if x[0]=='Y'])
+        overall_yes_perc = ((yes_sum * 100) / yes_no_sum) / 100
```
Comment (kilemensi) [inline reports/report_builder.py]: Why multiply by `100` then divide by `100`?

**Review by kilemensi — COMMENTED** [review]
The approach seems fine @esirK ... very _manual_ but I can't think of a better/faster way of doing it.

File: reports/report_builder.py [hunk 4]
```diff
@@ -1193,6 +1210,17 @@ def ws_28(self, ws):
                 secondary_counts[media_type] = counts
             self.tabulate_secondary_cols(ws, secondary_counts, self.male_female, self.regions, row_perc=True, show_N=True)
             self.tabulate_historical(ws, '28', self.male_female, self.regions, r=7)
+        overall_row = ws.dim_rowmax + 2
+        ws.write(overall_row, overall_column-1, "Overall", self.label)
+        overall_column +=1
+        for media_type in secondary_counts:
+            setattr(self, f"female_male_sum_{media_type}", sum(secondary_counts[media_type].values()))
+            setattr(self, f"female_sum_{media_type}", sum([secondary_counts[media_type][x] for x in secondary_counts[media_type] if x[0] in self.female_ids]))
+            value = getattr(self, f"female_sum_{media_type}")/getattr(self, f"female_male_sum_{media_type}")
```
Comment (kilemensi) [inline reports/report_builder.py]: Any reason why we need to set these in `self` i.e. why not just declare local variables?

File: reports/report_builder.py [hunk 5]
```diff
@@ -4646,6 +4758,47 @@ def write_primary_row_heading(self, ws, heading, c=0, r=6):
...
+        ws.write(r, c-1, "Overall Original Tweets", self.label)
+        ws.write(r, c+1, value, self.P)
+
+    def write_female_overall(self, ws, counts, c, r, write_overall=True):
+        male_female_sum = sum(counts.values())
+        female_sum = sum([counts[x] for x in counts if x[0] in self.female_ids])
+        if write_overall:
+            ws.write(r, c-1, "Overall Female", self.label)
+        value = female_sum/male_female_sum
+        ws.write(r, c+1, value, self.P)
```
Comment (kilemensi) [inline reports/report_builder.py]: I think we can reduce duplication in these write overall methods by introducing
```python
  def write_overall_value(self, ws, value, total, c, r, write_overall, overall_label = "Overall"):
        p_value = value / total
        if write_overall:
            ws.write(r, c-1, overall_label, self.label)
        ws.write(r, c+1, p_value, self.P)
```
Then all these methods will use ☝🏾

**Review by kilemensi — COMMENTED** [review]
👍🏾 @esirK ... I think technically it's ready

----------------------------

1. I think lets align overall value with the column it represents e.g. overall female % should be in the Female column as 👇🏾,
Screenshot from 2021-03-16 16-00-39
2. The `write_xxxxx_overall` methods should be renamed to `write_overall_xxxxx`

File: reports/report_builder.py (hunk 4 above)
Comment (esirK) [reply reports/report_builder.py]: I have no idea what I was thinking here. Fixing this now.

**Review by kilemensi — COMMENTED** [review]
👍🏾 @esirK ... Looks alright.

Any reason that the overall value is still displayed on the wrong column? For example, Sheet 5 👇🏾 , the I believe the overall is for female but it's been written on the male columns
Screenshot from 2021-03-17 09-25-55

**Comment by esirK** [comment]
> Any reason that the overall value is still displayed on the wrong column? For example, Sheet 5 👇🏾 , the I believe the overall is for female but it's been written on the male columns

This is fixed. Just that I didn't re-upload the screenshots. I've uploaded the new ones.

**Comment by kilemensi** [comment]
> > Any reason that the overall value is still displayed on the wrong column? For example, Sheet 5 👇🏾 , the I believe the overall is for female but it's been written on the male columns
> 
> This is fixed. Just that I didn't re-upload the screenshots. I've uploaded the new ones.

The screenshot was from my local test @esirK

**Comment by esirK** [comment]
That is odd @kilemensi  because if you look at [this](https://github.com/CodeForAfrica/gmmp/blob/ft-add-overall/reports/report_builder.py#L665) you'll see we aren't passing an `overall_label` so the default should be used. But from your screenshot, the overall label is a custom one. 
Maybe something to do with restarting docker?

**Review by kilemensi — APPROVED** [review]
🚀

**Pull request merged**